File: 1800-concatenation-of-consecutive-binary-numbers/concatenation-of-consecutive-binary-numbers.cpp
class Solution {
    int mod=1e9+7;
public:
    int concatenatedBinary(int n) {
        
        int powlen=0;
        long long ans=0;

        for(int i=1;i<=n;i++){
            if((i&(i-1))==0){
                powlen++;
            }
            ans=((ans<<powlen)%mod+i)%mod;
        }
        return ans;
    }
};

File: 1800-concatenation-of-consecutive-binary-numbers/concatenation-of-consecutive-binary-numbers.java
class Solution {
    int mod = 1000000000+7;
    public int concatenatedBinary(int n) {
        int pow2=1;

        int ans = 0;

        for(int i=n;i>0;i--){
            int curr = i;

            while(curr>0){
                int bit = ((curr)&(1));
                if(bit==1){
                    ans = (ans + pow2)%mod;
                }
                pow2 = (pow2%mod*2%mod)%mod;
                curr=curr>>1;
            }
        }

        return ans;
    }
}